src/framework/node0.ml

```diff
@@ -7,16 +7,22 @@ type t =
   | Statement of CilType.Stmt.t
   (** The statements as identified by CIL *)
   (* The stmt in a Statement node is misleading because nodes are program points between transfer functions (edges), which actually correspond to statement execution. *)
+  | LongjmpTo of CilType.Stmt.t
+  (** The statement as the target of a longjmp *)
   | FunctionEntry of CilType.Fundec.t
   (** *)
   | Function of CilType.Fundec.t
   (** The variable information associated with the function declaration. *)
+  | LongjmpFromFunction of CilType.Fundec.t
+  (** Handles longjumps out of the concerned function *)
```

Comment:
Wasn't the idea to do an analysis without using program transformations? Or did I get the wrong impression?

It seems to me that these dynamic things shouldn't require CFG modifications, but should be doable with their existing call nodes and edges:
* `setjmp` returning twice could just be a join of the two things within the call transfer function instead of requiring two incoming edges that get joined by the CFG shape.
* `longjmp` being an early return should also be doable because we already do thread returns and thread return values just fine.

One could be calling `setjmp`/`longjmp` via function pointers, so CFG transformation cannot be complete. Wasn't there even some real-world example where a custom `longjmp` could be supplied (as a function pointer)?

Reply:
This is not a CFG transformation, it just adds (unconnected!) new global(!) unknowns to the analysis (`ret'_f` is `LongjmpFromFunction`), `LongjmpTo` is needed because there can be no direct side-effect to unknowns that have a right-hand side due to a solver limitation.  Technically, this once again is a global queried during RHS evaluation, and not the CFG transformation splitting a `setjmp` into two different transfer functions as we describe in the paper. 

 I'm not sure if there is a better way to do this, as these unknowns need to live outside any particular analysis and outside any lifters, so I thought it might be best to add them to the CFG as unconnected nodes.

> setjmp returning twice could just be a join of the two things within the call transfer function instead of requiring two incoming edges that get joined by the CFG shape.

This is exactly what happens, though it has to happen at the top level, i.e., in `constraints.ml` and cannot happen in any particular analysis.

> longjmp being an early return should also be doable because we already do thread returns and thread return values just fine.

This also is done as a side-effect to an additional node. The new node is needed because this should not be specific to any particular analysis, so once again is handled in `constraints.ml`


> One could be calling setjmp/longjmp via function pointers, so CFG transformation cannot be complete. Wasn't there even some real-world example where a custom longjmp could be supplied (as a function pointer)?

Yes, there is also in the regression tests. Here, we first rely on `constraints.ml` resolving the potentially called functions. A call to `setjmp` or `longjmp` then is handled in `tf_special_call` as are any other calls to special functions via pointers.

Reply:
I see, I missed the fact that these aren't actually added into the CFG, but just defined in the CFG edge type. This is rather confusing though because all the CFG-handling code (e.g. Graphviz output) must handle constructs which never occur.

Arbitrary global constraint variables can be introduced as `Spec.V`. That not only includes individual analyses but also lifters and at the outer level here in `FromSpec` (which defines `GVar` of the constraint system from `Spec.V`). In fact, `FromSpec` already introduces a set of its own (write-only) constraint unknowns for some context lookups.
Something similar can be done for these extra things. It would be amazing if it could even be extracted to a lifter since that would avoid a 4-way `Either` of global constraint variables in `FromSpec`.

Reply:
Come to think of it, these should probably not be added to `Spec.V` but instead be part of `LVar`, as are the entry points and return points of functions, as they also take values from the lattice of locals, and not from some dedicated ones for globals.

Reply:
I looked at changing `VarF`, i.e. `LVar` but that is super ugly and propagates absolutely everywhere, so actually I now think making this modification to the Node type is the right call. These behave like nodes in every possible way, i.e., are enhanced with contexts to get unknowns taking values from `D.t`,  get side-effects via `sideg`, ...

So i think for this conceptual simplicity, I think having some cases where these two new types of nodes are to be ignored is a small price to pay.

Reply:
> Come to think of it, these should probably not be added to `Spec.V` but instead be part of `LVar`, as are the entry points and return points of functions, as they also take values from the lattice of locals, and not from some dedicated ones for globals.

At this point the distinction between `LVar` and `GVar` is really just historical. The customizability of `GVar` is already used in numerous places for similar things, e.g. as I mentioned somewhere thread returns. Also the dead branch lifter just induces per-`Node.t` unknowns for itself. We now use side-effects for a lot more than just function entry and global value invariants, so the distinction is already blurred and we've used them in a more general way.

What I really don't want is the implementation details of an analysis to leak through the abstractions we have maintained. I might want to have a look at changing this myself, but that can wait until after the paper submission. Paper-wise it doesn't have to change anything but keeping a separation of concerns in the implementation is valuable for all future maintenance and development.